### PR TITLE
HDDS-5263. SCM may stay in safe mode forever after a unclean shutdown of SCM.

### PR DESCRIPTION
**Pull request by bharatviswa504**
## What changes were proposed in this pull request?

After unclean SCM shutdown, SCM may not come out of safemode.

Attached a document to Jira with the problem statement and the proposal.

Proposal:
1. Use leader ready to start Background services.
2. In apply transaction after apply is complete if SCM is in safemode, refresh and validate safemode rules with current state.
3. For leader ready have a back ground daemon thread and check is leader ready using Ratis API. And also call notifyStatusChanged in SCMServices and also update isLeader in SCMContext.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-5263

## How was this patch tested?
Tested the fix on a cluster.
Testing is described here [link](https://github.com/apache/ozone/pull/2294#issuecomment-854422133)
Performed failover testing also to check leader status is properly propogated.
Added tests for SCMContext new API.

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/XceiverServerRatis.java [hunk 1]
```diff
@@ -437,7 +437,7 @@ private RpcType setRpcType(RaftProperties properties) {
 
   private void setPendingRequestsLimits(RaftProperties properties) {
 
-    final int pendingRequestsByteLimit = (int)conf.getStorageSize(
+    final long pendingRequestsByteLimit = (int)conf.getStorageSize(
```
Comment (bshashikant) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/XceiverServerRatis.java:440]: The change is not related. Let's revert this in this jira.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 2]
```diff
@@ -132,6 +135,17 @@ public void initialize(RaftServer server, RaftGroupId id,
       final SCMRatisRequest request = SCMRatisRequest.decode(
           Message.valueOf(trx.getStateMachineLogEntry().getLogData()));
       applyTransactionFuture.complete(process(request));
+      // After restart ratis replay logs from last snapshot index.
+      // So if some transactions which need to be updated to DB will not be
+      // applied to DB. After a restart of SCM container/pipeline managers
+      // have setup the safemode rules with not to update DB. Due to this
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:141]: The comment is not clear here.

Comment (bshashikant) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:141]: Its better to mention safeMode rules might work with stale data.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 3]
```diff
@@ -132,6 +135,17 @@ public void initialize(RaftServer server, RaftGroupId id,
...
+      // some time safemode rules are not validated and SCM does not exit
+      // safe mode. So, once after restart as transactions are applied, we
+      // check whether safe mode rules are validated to solve the issue of
+      // SCM not coming out of safemode.
+      if (scm.isInSafeMode()) {
+        scm.getScmSafeModeManager().refreshAndValidate();
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:147]: probably, this code will go off in the next patch.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 4]
```diff
@@ -80,6 +81,8 @@
   // and reinitialize().
   private DBCheckpoint installingDBCheckpoint = null;
 
+  private Daemon leaderReady;
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:84]: leaderReady -> leaderReadyDetector

Comment (bshashikant) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:84]: Please add some doc on what this thread is supposed to do.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/XceiverServerRatis.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/XceiverServerRatis.java]: Done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java (hunk 3 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:148]: Looks like we still need it. Have a look in to latest update

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java (hunk 4 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java]: Removed this

**Comment by bharatviswa504** [comment]
Tested this on cluster 
with scenario, close pipeline, scrubber removed and created new pipeline.

And once after leader ready, rules refreshed and data nodes registered, rules are successfully validated. (Unlike before after remove pipeline, pipeline rules are not successfully validated)

2021-06-04 07:05:28,646 INFO org.apache.hadoop.hdds.scm.safemode.HealthyPipelineSafeModeRule: Refreshed total pipeline count is 1, healthy pipeline threshold count is 1
2021-06-04 07:05:28,646 INFO org.apache.hadoop.hdds.scm.safemode.OneReplicaPipelineSafeModeRule: Refreshed Total pipeline count is 1, pipeline's with at least one datanode reported threshold count is 1

2021-06-04 07:05:30,098 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:31,146 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:32,009 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:32,059 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:33,841 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:33,888 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:33,921 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:39,331 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 0, required healthy pipeline reported count is 1
2021-06-04 07:05:39,332 INFO org.apache.hadoop.hdds.scm.safemode.SCMSafeModeManager: SCM in safe mode. Healthy pipelines reported count is 1, required healthy pipeline reported count is 1

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMContext.java [hunk 5]
```diff
@@ -96,6 +96,38 @@ public void updateLeaderAndTerm(boolean leader, long newTerm) {
     }
   }
 
+  /**
+   * Update isLeader flag.
+   * @param leader
+   */
+  public void updateLeader(boolean leader) {
+    lock.writeLock().lock();
+    try {
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMContext.java:105]: This looks a bit confusing. It should be ideally leaderReady. Let's maintain notion of leader and leaderReady separately.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 6]
```diff
@@ -297,6 +308,23 @@ public void notifyTermIndexUpdated(long term, long index) {
     // with some information like its peers and termIndex). So, calling
     // updateLastApplied updates lastAppliedTermIndex.
     updateLastAppliedTermIndex(term, index);
+
+    if (currentLeaderTerm.get() == term &&
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:309]: Can we add comments here on why the logic is necessary?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/HealthyPipelineSafeModeRule.java [hunk 7]
```diff
@@ -130,30 +124,61 @@ protected void process(Pipeline pipeline) {
...
+
   @Override
-  protected void cleanup() {
+  protected synchronized void cleanup() {
     processedPipelineIDs.clear();
   }
 
   @VisibleForTesting
-  public int getCurrentHealthyPipelineCount() {
+  public synchronized int getCurrentHealthyPipelineCount() {
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/HealthyPipelineSafeModeRule.java:168]: Why these calls need to be synchronized()?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMContext.java (hunk 5 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMContext.java:105]: Why do we need 2 leader and leader ready. Actually we need leader ready every where.
Do you see any use-case for both?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/HealthyPipelineSafeModeRule.java (hunk 7 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/HealthyPipelineSafeModeRule.java:168]: To avoid contention between refresh and process reports.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java (hunk 6 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:312]: Done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMContext.java (hunk 5 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMContext.java:115]: Updated to fix this confusion and also added 2 APIs.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 8]
```diff
@@ -297,6 +305,33 @@ public void notifyTermIndexUpdated(long term, long index) {
     // with some information like its peers and termIndex). So, calling
     // updateLastApplied updates lastAppliedTermIndex.
     updateLastAppliedTermIndex(term, index);
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:307]: Probably need to do an isInitlaized() check here, so that none of below code gets executed during scm init()?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 9]
```diff
@@ -297,6 +305,33 @@ public void notifyTermIndexUpdated(long term, long index) {
...
+      // This means after a restart, all pending transactions have been applied.
+      // Perform
+      // 1. Refresh Safemode rules state.
+      // 2. Start DN Rpc server.
+      if (!refreshedAfterLeaderReady.get()) {
+        scm.getScmSafeModeManager().refresh();
+        scm.getDatanodeProtocolServer().start();
+
+        refreshedAfterLeaderReady.set(true);
+      }
```
Comment (bshashikant) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:328]: Remove the empty lines.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java (hunk 8 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:312]: Done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java (hunk 9 above)
Comment (bharatviswa504) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:333]: Done

**Review by bshashikant — APPROVED** [review]

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java [hunk 10]
```diff
@@ -286,17 +294,49 @@ public long takeSnapshot() throws IOException {
...
+
+      // Means all transactions before this term have been applied.
+      // This means after a restart, all pending transactions have been applied.
+      // Perform
+      // 1. Refresh Safemode rules state.
+      // 2. Start DN Rpc server.
+      if (!refreshedAfterLeaderReady.get()) {
+        scm.getScmSafeModeManager().refresh();
+        LOG.info("bharat starting from sm");
```
Comment (vivekratnavel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java:333]: @bharatviswa504 I just noticed that this line got committed by mistake. Can we rectify this in a separate jira or combine it in your next patch?
